### ISSUE TITLE
Feature request: allow CLIEngine.executeOnFiles consumers to skip globbing

### ISSUE THREAD

**Issue by pvdlg** [eslint/eslint#9972]
The current implementation of [CLIEngine.executeOnFiles](https://github.com/eslint/eslint/blob/47ac478f3272fca771461e1cc6d4e91b1ecaaccc/lib/cli-engine.js#L491) accept an `Array` of glob patterns and [resolve them](https://github.com/eslint/eslint/blob/master/lib/cli-engine.js#L517). In certain situation API consumer might already know the exact list of file they want to lint and the globbing done by `eslint` could be skipped.

This is the case of [xojs/xo](https://github.com/xojs/xo) for example, where we [glob the files](https://github.com/xojs/xo/blob/652a6e51a8812e1d573c55159d8f55794e97b870/index.js#L92) to lint, based on predefined patterns and ignore. We have to determine the list of file to lint, rather than just passing glob patterns to `CLIEngine.executeOnFiles` in order to implement certain features specific to `xo`.

As a result the files are globbed twice, once by `xo`, then once by `eslint`. The globbing done by `eslint` is redundant in that case as the file paths to lint are already determined and the ones to ignore are already excluded.

The proposed feature would provide a way to call `CLIEngine.executeOnFiles` with a list of paths rather than a list of glob patterns. When called this way `eslint` would process the file paths passed, skipping the glob resolution. This would result in a performance increase for users of tools consuming the `eslint` API.

I'm mentioning the case of `xo` as an example, but I believe it could be useful for other tools consuming the `CLIEngine.executeOnFiles` API.

As far as my test went, I think the change would be fairly simple, backward compatible and non obstructive for `eslint`.

`CLIEngine.executeOnFiles` could accept a second parameter, `filePaths` and
```js
const fileList = globUtil.listFilesToProcess(this.resolveFileGlobPatterns(patterns), options)
```
would become something along those lines:
```js
const fileList = (filePaths || []).map(filePath => ({filename: filePath, ignored: false})) .concat(globUtil.listFilesToProcess(this.resolveFileGlobPatterns(patterns), options));
``` 

Or instead `CLIEngine.executeOnFiles` could accept either:
- an `Array` in which case it would behave as of now
- and Object with a `patterns` and a `filePaths` property in which case the `patterns` would be handled as of now, and the `filePaths` as proposed above

In my very rough tests done on a test project with large amount of files the performance gain in the case of `xo` is around a 20% reduction in running time.

**Comment by not-an-aardvark** [comment]
Hi, thanks for the proposal. I'm trying to understand the issue -- would it work to just pass a list of filepaths to `executeOnFiles` right now? The paths would be interpreted as globs, but since the paths wouldn't have any wildcards, it seems like there would only be a small amount of overhead in comparison to re-resolving the original glob patterns.

**Comment by pvdlg** [comment]
Hi!

> would it work to just pass a list of filepaths to executeOnFiles right now? The paths would be interpreted as globs, but since the paths wouldn't have any wildcards, it seems like there would only be a small amount of overhead in comparison to re-resolving the original glob patterns.

This is what happens now and it works perfectly. However passing a list of file paths (without wildcards) to `glob`is certainly faster than passing a list of glob with wildcards.

But the overhead of globbing the file paths without wildcards is non neglectable.

In my tests the time spend by `globUtil.listFilesToProcess(this.resolveFileGlobPatterns(patterns), options)` to go through those file paths without wildcard is roughly 20% of the total time (which includes our globbing + eslint "redundant" globbing + rule execution).
I don't have the exact number, but the globbing of those files paths without wildcard represent a large part of the total eslint processing time.

If you'd like I can create a sample reproduction project, calling `executeOnFiles` with a lot of file paths and measure the time spent by `globUtil.listFilesToProcess(this.resolveFileGlobPatterns(patterns), options)`.

**Comment by pvdlg** [comment]
I'm using this repo to test: https://github.com/pixelass/ava-xo-test. It's an unrealistically big project, so it allow to better measure perf by limiting the inherent variability in perf of the FS.

When "pre-globbing" and passing a list of file paths without wildcards to `executeOnFiles` I get:
- Total `executeOnFiles` time: 11 208ms
- Time to execute `globUtil.listFilesToProcess(this.resolveFileGlobPatterns(patterns), options)`: 3998ms

So 35% of the total time executing`executeOnFiles` is spent globbing file paths without wildcards.

Let me know if you'd like me to make this test available in a way that is easy for you to run.

**Comment by sindresorhus** [comment]
I think a `glob: false` option would make the most sense, and I've seen that in other modules that accept globs. Not just for performance reasons, but also if you need to make sure paths are handled raw. Globbing has a lot of weird rules and syntax. Imagine you want to allow file paths containing parens.

**Comment by not-an-aardvark** [comment]
Sounds reasonable to me.

**Comment by pvdlg** [comment]
Awesome!!! I'll work on the PR ASAP with `glob: false` as suggested by @sindresorhus

**Comment by not-an-aardvark** [comment]
Since this would be a semver-minor change to ESLint core, in order to accept this it would need to be discussed on the agenda at the next TSC meeting (which is in two weeks). It's possible that people will suggest changes to the proposal before then or at the meeting, so while you're certainly free to create a PR now, you might reduce the risk of accidentally wasting your time if you wait until the proposal is accepted.

**Comment by pvdlg** [comment]
Thanks for letting me know! Very much appreciated!
I will not rush to make the PR then and do when I have some free time or if/when it's accepted.

**Comment by ilyavolodin** [comment]
**TSC Description:** Add `glob: false` to `CLIEngine.executeOnFiles` to avoid additional overhead when passing in full file paths.

**TSC Question:** Should this change be accepted?

**Comment by platinumazure** [comment]
Question: Should we have any special handling for `glob: false` and a path comes in that doesn't point to a file, but might "look like" a glob? Should we consider printing out a warning in that case?

Should we consider adding a CLI option as well? (I know this proposal doesn't include it at this point, and they could be added independently, so I'm not trying to weigh down this proposal with extra stuff. Just curious about the team's or community's thoughts.

**Comment by sindresorhus** [comment]
> Should we have any special handling for glob: false and a path comes in that doesn't point to a file, but might "look like" a glob? Should we consider printing out a warning in that case?

No, it's explicitly set to no globbing. This is an API for developers, not end-users.

**Comment by not-an-aardvark** [comment]
This issue was accepted in [today's TSC meeting](https://github.com/eslint/tsc-meetings/blob/db3c0010144ba403a95daf3611af24ee271b3c78/notes/2018/2018-03-01.md#feature-request-allow-cliengineexecuteonfiles-consumers-to-skip-globbing). However, there was consensus that the name of the option should be `globInputPaths` rather than `glob`, to clarify that other types of globs (e.g. `overrides` in config files, ignore patterns) are still resolved in either case.

**Comment by pvdlg** [comment]
Awesome! I'll start working on the implementation with a `globInputPaths` option.

**Comment by platinumazure** [comment]
Hi @pvdlg, friendly ping: Are you still going to work on an implementation? No rush, just wanted to make sure we knew where things were at here :smile:

**Comment by pvdlg** [comment]
Hi yes I will work on the implementation. I had unforeseen personal stuff to deal with which limited my open source time.
I don't know when exactly when I'll have time unfortunately but it's definitely in my todo list :)
